File: backend/engine/fund_returns.py
"""Fund-level (LP-facing) return overlay — refactor plan P4-2 (Python mirror).

Mirrors lib/engine/fundReturns.ts. Translates the deal's gross equity cashflow
stream (returns.equity_cashflows) into NET LP returns after management fees and
GP carried interest over a preferred return. Pure overlay — never touches the
core deal engine, so deal-level IRR/MOIC are unchanged.

Documented simplifications (kept deliberately simple and defensible, single-deal
scope): management fees are a per-year drag (years 1..hp); net MOIC = net value /
invested capital; preferred return compounds annually; European = carry at exit
on total profit above return-of-capital + pref; American = carry as distributions
clear the running capital + accrued-pref high-water; no GP catch-up, no clawback.
"""

from __future__ import annotations

from typing import Optional

from backend.models.outputs import FundReturns, Returns
from backend.models.state import ModelState
from backend.engine.returns import _build_time_vector, _solve_irr_auto


def compute_fund_returns(state: ModelState, returns: Returns) -> Optional[FundReturns]:
    fa = state.fund_assumptions
    if fa is None:
        return None

    hp = state.exit.holding_period
    cfs = returns.equity_cashflows or []
    invested = returns.entry_equity
    gross_irr = returns.irr
    gross_moic = returns.moic

    if invested <= 0 or len(cfs) < 2:
        return FundReturns(
            net_irr=None, net_moic=0.0, gross_irr=gross_irr, gross_moic=gross_moic,
            gross_to_net_spread=None, management_fees_total=0.0, carried_interest=0.0,
            preferred_return_shortfall=0.0, lp_paid_in=max(0.0, invested), lp_distributions=0.0,
        )

    gross_inflows = [cfs[i] if i < len(cfs) else 0.0 for i in range(1, hp + 1)]
    total_receipts = sum(gross_inflows)

    # ── Management fee (annual drag, years 1..hp) ──
    if fa.management_fee_basis == "committed":
        fee_basis = fa.fund_size * min(1.0, max(0.0, fa.deal_allocation_pct))
    else:
        fee_basis = invested
    fee_per_year = max(0.0, fa.management_fee_pct) * max(0.0, fee_basis)
    fee_by_year = [fee_per_year] * hp
    management_fees_total = sum(fee_by_year)

    # ── Carried interest ──
    pref = max(0.0, fa.preferred_return)
    carry_rate = max(0.0, fa.carry_rate)
    carry_by_year = [0.0] * hp
    carried_interest = 0.0
    preferred_shortfall = 0.0

    if fa.carry_waterfall == "american":
        remaining_capital = invested
        accrued_pref = 0.0
        for i in range(hp):
            accrued_pref += remaining_capital * pref
            receipt = gross_inflows[i]
            cap_return = min(receipt, remaining_capital)
            remaining_capital -= cap_return
            receipt -= cap_return
            pref_paid = min(receipt, accrued_pref)
            accrued_pref -= pref_paid
            receipt -= pref_paid
            if receipt > 0:
                carry_by_year[i] = carry_rate * receipt
                carried_interest += carry_by_year[i]
        preferred_shortfall = accrued_pref if remaining_capital > 1e-9 else 0.0
    else:
        pref_amt = invested * ((1.0 + pref) ** hp - 1.0)
        profit_above_pref = max(0.0, total_receipts - invested - pref_amt)
        carried_interest = carry_rate * profit_above_pref
        carry_by_year[hp - 1] = carried_interest
        capital_returned = min(total_receipts, invested)
        pref_met = max(0.0, total_receipts - capital_returned)
        preferred_shortfall = max(0.0, pref_amt - pref_met)

    # ── Net LP cashflows ──
    times = _build_time_vector(hp, state.exit.mid_year_convention) if state.exit.mid_year_convention else None
    lp_cfs = [-invested]
    for i in range(hp):
        lp_cfs.append(gross_inflows[i] - fee_by_year[i] - carry_by_year[i])
    net_irr = _solve_irr_auto(lp_cfs, times)

    lp_distributions = total_receipts - management_fees_total - carried_interest
    net_moic = lp_distributions / invested if invested > 0 else 0.0
    gross_to_net_spread = (gross_irr - net_irr) if (gross_irr is not None and net_irr is not None) else None

    return FundReturns(
        net_irr=net_irr,
        net_moic=net_moic,
        gross_irr=gross_irr,
        gross_moic=gross_moic,
        gross_to_net_spread=gross_to_net_spread,
        management_fees_total=management_fees_total,
        carried_interest=carried_interest,
        preferred_return_shortfall=preferred_shortfall,
        lp_paid_in=invested,
        lp_distributions=lp_distributions,
    )

File: lib/engine/fundReturns.ts
/**
 * Fund-level (LP-facing) return overlay — refactor plan P4-2.
 *
 * Translates the deal's gross equity cashflow stream into NET LP returns after
 * (1) management fees and (2) GP carried interest over a preferred return. This
 * is a clean overlay on `returns.equity_cashflows` — it never touches the core
 * deal engine, so deal-level IRR/MOIC are unchanged.
 *
 * Modelling conventions (documented simplifications — kept deliberately simple
 * and defensible rather than a full fund waterfall):
 *   • Management fees are a per-year drag on the equity cashflow (years 1..hp).
 *     Basis = invested capital (deal equity) or committed (fund_size × deal share).
 *   • Net MOIC = net value to LPs / invested capital (net multiple on invested).
 *   • Preferred return accrues on capital OUTSTANDING each year (time-weighted),
 *     not as a flat compounded amount on full capital for the whole hold.
 *   • Full waterfall per distribution: return of capital → preferred → 100% GP
 *     catch-up to carry% of (pref + catch-up) → carry% of the residual.
 *   • European (whole-fund): carry crystallised at exit. American (deal-by-deal):
 *     carry taken as distributions clear each tier.
 *   • Paid-in capital includes follow-on (add-on) equity; interim negative flows
 *     are capital calls, not carry-eligible distributions.
 *   • No clawback (single-deal scope).
 */

import type { ModelState, Returns, FundReturns, FundAssumptions } from '../dealEngineTypes';
import { solveIrr, solveIrrTimed } from './returns';

function fundTimeVector(hp: number, midYear: boolean): number[] | null {
  if (!midYear) return null;
  return [0, ...Array.from({ length: hp - 1 }, (_, t) => t + 0.5), hp];
}

function solveIrrAuto(cfs: number[], times: number[] | null): number | null {
  return times ? solveIrrTimed(cfs, times) : solveIrr(cfs);
}

export function computeFundReturns(state: ModelState, returns: Returns): FundReturns | undefined {
  const fa: FundAssumptions | undefined = state.fund_assumptions;
  if (!fa) return undefined;

  const hp = state.exit.holding_period;
  const cfs = returns.equity_cashflows ?? [];
  const invested = returns.entry_equity;
  const grossIrr = returns.irr;
  const grossMoic = returns.moic;

  // Degenerate deal (no solvable equity stream): return a zeroed overlay.
  if (invested <= 0 || cfs.length < 2) {
    return {
      net_irr: null, net_moic: 0, gross_irr: grossIrr, gross_moic: grossMoic,
      gross_to_net_spread: null, management_fees_total: 0, carried_interest: 0,
      preferred_return_shortfall: 0, lp_paid_in: Math.max(0, invested), lp_distributions: 0,
    };
  }

  // Gross inflows by hold year (index 0 = year 1 … hp-1 = year hp). A NEGATIVE entry is an
  // interim CAPITAL CALL (e.g. an equity-funded add-on draw), not a distribution — it adds
  // to paid-in capital, not to carry-eligible proceeds.
  const grossInflows: number[] = [];
  for (let i = 1; i <= hp; i++) grossInflows.push(cfs[i] ?? 0);

  // Total LP paid-in capital = entry equity + follow-on (add-on) equity called over the hold
  // (Bug 6: the overlay previously ignored follow-on capital, understating paid-in).
  const addOnEquity = returns.add_on_equity_invested ?? 0;
  const paidIn = invested + addOnEquity;

  // ── Management fee (annual drag, years 1..hp) ──
  const feeBasis = fa.management_fee_basis === 'committed'
    ? fa.fund_size * Math.max(0, Math.min(1, fa.deal_allocation_pct))
    : paidIn; // invested basis now includes follow-on capital
  const feePerYear = Math.max(0, fa.management_fee_pct) * Math.max(0, feeBasis);
  const feeByYear = grossInflows.map(() => feePerYear);
  const managementFeesTotal = feeByYear.reduce((s, f) => s + f, 0);

  // ── Carry waterfall (time-weighted, with 100% GP catch-up) ──
  // Per positive distribution: (1) return of capital, (2) preferred return accrued on
  // capital OUTSTANDING (time-weighted, not a flat compounded amount), (3) GP catch-up to
  // carry% of (pref + catch-up), (4) carry% of the residual. American crystallises carry as
  // distributions clear; European defers all carry to exit. Interim negative flows raise
  // outstanding capital (a capital call), never trigger carry.
  const pref = Math.max(0, fa.preferred_return);
  const carryRate = Math.max(0, Math.min(0.99, fa.carry_rate));
  const european = fa.carry_waterfall !== 'american';
  const carryByYear = new Array<number>(hp).fill(0);
  let remainingCapital = invested;
  let accruedPref = 0;
  let cumPrefPaid = 0;
  let cumCarry = 0;
  let grossDistributions = 0;
  for (let i = 0; i < hp; i++) {
    accruedPref += remainingCapital * pref; // accrue on capital outstanding this year
    let receipt = grossInflows[i];
    if (receipt < 0) { remainingCapital += -receipt; continue; } // interim capital call
    grossDistributions += receipt;
    // (1) return of capital
    const capReturn = Math.min(receipt, remainingCapital);
    remainingCapital -= capReturn; receipt -= capReturn;
    // (2) preferred return
    const prefPaid = Math.min(receipt, accruedPref);
    accruedPref -= prefPaid; cumPrefPaid += prefPaid; receipt -= prefPaid;
    // (3) GP catch-up to carry% of (pref + catch-up): target cumCarry = r/(1−r) × cumPrefPaid
    const catchUpTarget = carryRate < 1 ? (carryRate / (1 - carryRate)) * cumPrefPaid : Infinity;
    const catchUp = Math.max(0, Math.min(receipt, catchUpTarget - cumCarry));
    cumCarry += catchUp; receipt -= catchUp;
    // (4) residual carry split
    const split = carryRate * receipt;
    cumCarry += split; receipt -= split;
    if (!european) carryByYear[i] = catchUp + split; // American: pay as you go
  }
  const carriedInterest = cumCarry;
  if (european) carryByYear[hp - 1] = carriedInterest; // European: all carry crystallised at exit
  // Preferred shortfall = pref still accrued and unpaid at the end (time-weighted).
  const preferredShortfall = Math.max(0, accruedPref);

  // ── Net LP cashflows ──
  const times = fundTimeVector(hp, state.exit.mid_year_convention ?? false);
  const lpCfs: number[] = [-invested];
  for (let i = 0; i < hp; i++) {
    lpCfs.push(grossInflows[i] - feeByYear[i] - carryByYear[i]);
  }
  const netIrr = solveIrrAuto(lpCfs, times);

  // Distributions to LPs = positive proceeds less fees and carry, over total paid-in.
  const lpDistributions = grossDistributions - managementFeesTotal - carriedInterest;
  const netMoic = paidIn > 0 ? lpDistributions / paidIn : 0;
  const grossToNetSpread = grossIrr != null && netIrr != null ? grossIrr - netIrr : null;

  return {
    net_irr: netIrr,
    net_moic: netMoic,
    gross_irr: grossIrr,
    gross_moic: grossMoic,
    gross_to_net_spread: grossToNetSpread,
    management_fees_total: managementFeesTotal,
    carried_interest: carriedInterest,
    preferred_return_shortfall: preferredShortfall,
    lp_paid_in: paidIn,
    lp_distributions: lpDistributions,
  };
}
